src/Data/List/Base.agda

```diff
@@ -49,6 +49,14 @@ map : (A → B) → List A → List B
 map f []       = []
 map f (x ∷ xs) = f x ∷ map f xs
 
+catMaybes′ : List (Maybe A) → List A
+catMaybes′ []             = []
+catMaybes′ (nothing ∷ xs) = catMaybes′ xs
+catMaybes′ (just x  ∷ xs) = x ∷ catMaybes′ xs
```

Comment:
```Agda
catMaybes′ (x  ∷ xs) = maybe _∷_ id x $ catMaybes′ xs
```
?

Reply:
Much better!

Reply:
What about
```agda
catMaybes″ : List (Maybe A) → List A
catMaybes″ = foldr (maybe′ _∷_ id) []
```
(starting to get dizzy with the possibilities, once you start 'fixing' things...)

Reply:
Personally that is my favourite because it makes reasoning about it simplest: it is very clear which lemmas will apply.

The downside is that it's not very pedagogical, but I don't think that's something we should be aiming for. [More precisely: I would be happy with a `Data.List.Traditional` module with the expanded definitions and proofs of equivalence to the less-expanded definitions used in `.Base`.

Reply:
Second point might be worth discussing separately as an issue (elsewhere people have complained to me about the lack of a(n equivalent to `haskell`) 'prelude' with which to teach (no `Level`s, 'elementary' definitions, ...)